### ISSUE TITLE
allow_new_users setting

### ISSUE THREAD

**Issue by jakobkukla** [mozilla-services/syncstorage-rs#1429]
With the python2 `syncserver` implementation it was possible to disable new user requests as seen below.

```
[syncserver]
...other settings...
allow_new_users = false
```

Is there an equivalent setting in `syncstorage-rs`? I couldn't find anything fitting while looking through the code but it is possible I missed it.

Thanks!

**Comment by jrconlin** [comment]
No, I don't believe that this is available in the rust version.

**Comment by HorlogeSkynet** [comment]
Hi !

This feature is definitely missing for users coming from the legacy `syncserver`, so I second this issue (as well as #1484, #1566 and #1629).

Could you tell us how you want to address it from a (software) architectural point of view, so one could try to work on and propose an implementation ?

Thanks for your work, bye 👋

**Comment by pjenvey** [comment]
@HorlogeSkynet it should work similarly to the old Python Tokenserver implementation of it: (where it occurs is here https://github.com/mozilla-services/tokenserver/blob/9128c9db356601696f91c6f1590253b44f99aafe/tokenserver/views.py#L372)

You'll want to:

- Add a new setting, e.g. `SYNC_TOKENSERVER__ALLOW_NEW_USERS` defaulting to `true`
- Plug in the logic that checks this setting. This is a little more involved in the Rust version, as in the Python code the `get_user` / \<check allow_new_users setting\> / `create_user` logic is all in the web handler/view, whereas in Rust this code ends up being in the `TokenserverDb::get_or_create_user` method. I think simply adding an `allow_new_users` field to this method's `GetOrCreateUser` parameter would be fine. When rejecting new user creation this should return a new variant of Tokenserver's `DbError` that would be translated into a `TokenserverError` of unauthorized with a status of `new-users-disabled`
- Ideally adding a new test somewhere to verify new users can't be created. The python version has one that makes a web request, but we don't have a lot of existing tests like these in the rust version that run against tokenserver (only really syncstorage in tests.rs). An test against the TokenserverRequest extractor might be a little easier (and enough)

**Comment by HorlogeSkynet** [comment]
Thanks for your detailed answer @pjenvey, please see #2304 for implementation proposal.
